### PR TITLE
tailscale: add `dns_split_nameservers` resource

### PR DESCRIPTION
**Pull request by mpminardi**
Add `resource_dns_split_nameservers` to allow for controlling split DNS settings for a given tailnet.

Updates https://github.com/tailscale/corp/issues/19483

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.33.0
 	github.com/stretchr/testify v1.9.0
 	github.com/tailscale/hujson v0.0.0-20221223112325-20486734a56a
-	github.com/tailscale/tailscale-client-go v1.16.0
+	github.com/tailscale/tailscale-client-go v1.16.1-0.20240424180809-5a248d9547d4
```
Comment (mpminardi) [inline go.mod:12]: TODO:
- [ ] uprev to actual version once https://github.com/tailscale/tailscale-client-go/pull/78 is through and a new release is cut

File: docs/resources/dns_split_nameservers.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "tailscale_dns_split_nameservers Resource - terraform-provider-tailscale"
+subcategory: ""
+description: |-
+  The dnssplitnameservers resource allows you to configure split DNS nameservers for your Tailscale network. See https://tailscale.com/kb/1054/dns for more information.
```
Comment (mpminardi) [inline docs/resources/dns_split_nameservers.md:6]: Tried to snake case this to match the actual resource name but the autogen tool / CI didn't like that so 🤷

File: tailscale/resource_dns_split_nameservers.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/tailscale/tailscale-client-go/tailscale"
+)
+
+func resourceDNSSplitNameservers() *schema.Resource {
+	return &schema.Resource{
```
Comment (knyar) [inline tailscale/resource_dns_split_nameservers.go:13]: I'd suggest adding an importer as well to allow users to import existing Split DNS configuration to their Terraform state. I suspect the usual `&schema.ResourceImporter{StateContext: schema.ImportStatePassthroughContext}` would work here.

File: tailscale/resource_dns_split_nameservers.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func resourceSplitDNSNameserversRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*tailscale.Client)
+	splitDNS, err := client.DNSSplitDNS(ctx)
+	if err != nil {
+		return diagnosticsError(err, "Failed to fetch split DNS configs")
+	}
+
+	nameservers := splitDNS[d.Get("domain").(string)]
```
Comment (knyar) [inline tailscale/resource_dns_split_nameservers.go:44]: We might need something like this to consider the resource deleted when there are no split DNS servers configured for a given domain.

```suggestion
	nameservers, ok := splitDNS[d.Get("domain").(string)]
	if !ok {
	  d.SetId("")
	  return
	}
```

File: tailscale/resource_dns_split_nameservers.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	d.SetId(domain)
+	return nil
+}
+
+func resourceSplitDNSNameserversUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	if !d.HasChange("nameservers") {
+		return resourceSplitDNSNameserversRead(ctx, d, m)
+	}
+
+	client := m.(*tailscale.Client)
```
Comment (knyar) [inline tailscale/resource_dns_split_nameservers.go:80]: Would it make sense to just call `resourceSplitDNSNameserversCreate` here?

Comment (mpminardi) [reply tailscale/resource_dns_split_nameservers.go:80]: Good call, will just delegate off to create.

File: tailscale/resource_dns_split_nameservers.go (hunk 4 above)
Comment (mpminardi) [reply tailscale/resource_dns_split_nameservers.go:44]: Digging into this a bit more but making this change was breaking import.

Comment (mpminardi) [reply tailscale/resource_dns_split_nameservers.go]: Ended up changing to use `d.Id()` as this is what was breaking the import and is mentioned as necessary [here](https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/import#importer-state-function) if using `schema.ImportStatePassthrough`.

This seems to operate fine without the `ok` check in the scenarios I've tested (namely creation, deletion, import, update of split DNS nameservers) so I've left it out for now.

**Review by knyar — APPROVED** [review]

**Comment by aiell0** [comment]
Just chiming in to say I love using Tailscale and this is a feature I have been anticipating for a while! Good work @mpminardi

**Pull request merged**